oximeter/db/src/native/block.rs

```diff
@@ -0,0 +1,699 @@
+                ValueArray::DateTime64 { precision: *precision, values: vec![] }
+            }
+            DataType::Enum8(variants) => {
+                ValueArray::Enum8 { variants: variants.clone(), values: vec![] }
+            }
+            DataType::Nullable(inner) => ValueArray::Nullable {
+                is_null: vec![],
+                values: Box::new(ValueArray::empty(inner)),
+            },
+            DataType::Array(inner) => {
+                let inner_type = (**inner).clone();
+                ValueArray::Array { inner_type, values: vec![] }
+            }
+        }
+    }
+
+    /// Concatenate another value array to this.
+    ///
+    /// # Panics
+    ///
+    /// This panics if the two value arrays do not have the same types.
+    fn concat(&mut self, rhs: ValueArray) {
+        match (self, rhs) {
+            (ValueArray::UInt8(us), ValueArray::UInt8(mut them)) => {
+                us.append(&mut them)
+            }
+            (ValueArray::UInt16(us), ValueArray::UInt16(mut them)) => {
+                us.append(&mut them)
+            }
+            (ValueArray::UInt32(us), ValueArray::UInt32(mut them)) => {
+                us.append(&mut them)
+            }
+            (ValueArray::UInt64(us), ValueArray::UInt64(mut them)) => {
+                us.append(&mut them)
+            }
+            (ValueArray::UInt128(us), ValueArray::UInt128(mut them)) => {
+                us.append(&mut them)
+            }
+            (ValueArray::Int8(us), ValueArray::Int8(mut them)) => {
+                us.append(&mut them)
+            }
+            (ValueArray::Int16(us), ValueArray::Int16(mut them)) => {
+                us.append(&mut them)
+            }
+            (ValueArray::Int32(us), ValueArray::Int32(mut them)) => {
+                us.append(&mut them)
+            }
+            (ValueArray::Int64(us), ValueArray::Int64(mut them)) => {
+                us.append(&mut them)
+            }
+            (ValueArray::Int128(us), ValueArray::Int128(mut them)) => {
+                us.append(&mut them)
+            }
+            (ValueArray::Float32(us), ValueArray::Float32(mut them)) => {
+                us.append(&mut them)
+            }
+            (ValueArray::Float64(us), ValueArray::Float64(mut them)) => {
+                us.append(&mut them)
+            }
+            (ValueArray::String(us), ValueArray::String(mut them)) => {
+                us.append(&mut them)
+            }
+            (ValueArray::Uuid(us), ValueArray::Uuid(mut them)) => {
+                us.append(&mut them)
+            }
+            (ValueArray::Ipv4(us), ValueArray::Ipv4(mut them)) => {
+                us.append(&mut them)
+            }
+            (ValueArray::Ipv6(us), ValueArray::Ipv6(mut them)) => {
+                us.append(&mut them)
+            }
+            (ValueArray::DateTime(us), ValueArray::DateTime(mut them)) => {
+                us.append(&mut them)
+            }
```

Comment:
it occurs to me that this (and many of the other methods on this type) could be generated with help from a macro...not that you *need* to do that since you've clearly already done the work of writing this all out.